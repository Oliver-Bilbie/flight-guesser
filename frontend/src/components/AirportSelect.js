import React, { useEffect, useState } from "react";
import { Box, Select } from "grommet";

const AirportSelect = ({ setAirport }) => {
  const [options, setOptions] = useState([]);
  const [filteredOptions, setFilteredOptions] = useState([]);
  const [search, setSearch] = useState(false);

  useEffect(() => {
    let request = new XMLHttpRequest();
    const path = "https://nhqos29571.execute-api.eu-west-1.amazonaws.com";

    request.onerror = function () {
      setOptions(["Failed to load"]);
    };
    request.ontimeout = function () {
      setOptions(["Failed to load"]);
    };
    request.onload = function () {
      if (request.status === 200) {
        if (request.response.status === 200) {
          setOptions(request.response.response);
        } else {
          setOptions(["Failed to load"]);
        }
      } else {
        setOptions(["Failed to load"]);
      }
    };

    request.timeout = 10000;
    request.responseType = "json";
    request.open("GET", path);
    request.send();
  }, []);

  const handleSearch = (text) => {
    if (text === "") {
      setFilteredOptions([]);
      setSearch(false);
    } else {
      const exp = new RegExp(text, "i");
      setFilteredOptions(options.filter((o) => exp.test(o)));
      setSearch(true);
    }
  };

  const handleChange = (value) => {
    setAirport(value);
    setFilteredOptions([]);
    setSearch(false);
  };

  return (
    <Box direction="row" align="center">
      <Select
        options={search ? filteredOptions : options}
        onChange={(event) => handleChange(event.value)}
        onSearch={(text) => handleSearch(text)}
      />
    </Box>
  );
};

export default AirportSelect;
