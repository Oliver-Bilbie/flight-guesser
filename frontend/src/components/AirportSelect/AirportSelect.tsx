import React, { useState } from "react";
import { Box, Select, Text, Spinner } from "grommet";

interface AirportSelectProps {
  airports: string[];
  setSelection: (selection: string) => void;
}

const AirportSelect: React.FC<AirportSelectProps> = ({
  airports,
  setSelection,
}): React.ReactElement => {
  const [filteredOptions, setFilteredOptions] = useState(airports);

  const handleSearch = (text): void => {
    const exp = new RegExp(text, "i");
    setFilteredOptions(airports.filter((o) => exp.test(o)));
  };

  return (
    <Box gap="small">
      <Text>Destination:</Text>
      <Box align="center">
        {airports.length > 1 ? (
          <Select
            options={filteredOptions}
            onChange={(event): void => setSelection(event.value)}
            onSearch={(text): void => handleSearch(text)}
            placeholder="Select"
            searchPlaceholder="Search..."
          />
        ) : airports.length === 0 ? (
          <Spinner a11yTitle="loading" />
        ) : (
          <Text>{airports[0]}</Text>
        )}
      </Box>
    </Box>
  );
};

export default AirportSelect;
