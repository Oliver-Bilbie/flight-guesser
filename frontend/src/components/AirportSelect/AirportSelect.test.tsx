import React from "react";
import renderer from "react-test-renderer";
import "@testing-library/jest-dom";
import { render, screen, cleanup, fireEvent } from "@testing-library/react";
import AirportSelect from "./AirportSelect";

const test_airports = ["airport 1", "airport 2", "airport 3"];
const doNothing = (): void => void afterEach(cleanup);

test("Matches snapshot", () => {
  const tree = renderer
    .create(<AirportSelect airports={[]} setSelection={doNothing} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("Test for no airports", async () => {
  render(<AirportSelect airports={[]} setSelection={doNothing} />);

  // Check that the expected components are rendered
  expect(screen.queryByText("Destination:")).toBeInTheDocument();
  expect(screen.getByLabelText("loading")).toBeInTheDocument();
});

test("Test for error message", async () => {
  render(
    <AirportSelect airports={["Error Message"]} setSelection={doNothing} />
  );

  // Check that the expected components are rendered
  expect(screen.queryByText("Destination:")).toBeInTheDocument();
  expect(screen.queryByText("Error Message")).toBeInTheDocument();
});

test("Test for provided data", async () => {
  window.scrollTo = jest.fn(); // prevents jest error for unsupported function

  render(<AirportSelect airports={test_airports} setSelection={doNothing} />);

  // Check that the expected components are rendered
  expect(screen.queryByText("Destination:")).toBeInTheDocument();
  expect(screen.queryByPlaceholderText("Select")).toBeInTheDocument();
});

test("Test data selection", async () => {
  window.scrollTo = jest.fn(); // prevents jest error for unsupported function
  const mockSelection = jest.fn();

  render(
    <AirportSelect airports={test_airports} setSelection={mockSelection} />
  );

  // Test that the provided elements are available for selection
  fireEvent.click(screen.getByPlaceholderText("Select"));
  expect(screen.queryByText("airport 1")).toBeInTheDocument();
  expect(screen.queryByText("airport 2")).toBeInTheDocument();
  expect(screen.queryByText("airport 3")).toBeInTheDocument();

  // Test element selection
  fireEvent.click(screen.getByText("airport 1"));
  fireEvent.click(screen.getByPlaceholderText("Select"));
  fireEvent.click(screen.getByText("airport 2"));
  fireEvent.click(screen.getByPlaceholderText("Select"));
  fireEvent.click(screen.getByText("airport 3"));
  expect(mockSelection).toHaveBeenNthCalledWith(1, "airport 1");
  expect(mockSelection).toHaveBeenNthCalledWith(2, "airport 2");
  expect(mockSelection).toHaveBeenNthCalledWith(3, "airport 3");
});

test("Test data search for '1'", async () => {
  window.scrollTo = jest.fn(); // prevents jest error for unsupported function

  render(<AirportSelect airports={test_airports} setSelection={doNothing} />);

  // Test that the provided elements are available for selection
  fireEvent.click(screen.getByPlaceholderText("Select"));
  expect(screen.queryByText("airport 1")).toBeInTheDocument();
  expect(screen.queryByText("airport 2")).toBeInTheDocument();
  expect(screen.queryByText("airport 3")).toBeInTheDocument();

  // Test search for "1"
  fireEvent.change(screen.getByPlaceholderText("Search..."), {
    target: { value: "1" },
  });
  expect(screen.queryByText("airport 1")).toBeInTheDocument();
  expect(screen.queryByText("airport 2")).not.toBeInTheDocument();
  expect(screen.queryByText("airport 3")).not.toBeInTheDocument();
});

test("Test data search for '2'", async () => {
  window.scrollTo = jest.fn(); // prevents jest error for unsupported function

  render(<AirportSelect airports={test_airports} setSelection={doNothing} />);

  // Test that the provided elements are available for selection
  fireEvent.click(screen.getByPlaceholderText("Select"));
  expect(screen.queryByText("airport 1")).toBeInTheDocument();
  expect(screen.queryByText("airport 2")).toBeInTheDocument();
  expect(screen.queryByText("airport 3")).toBeInTheDocument();

  // Test search for "2"
  fireEvent.change(screen.getByPlaceholderText("Search..."), {
    target: { value: "2" },
  });
  expect(screen.queryByText("airport 1")).not.toBeInTheDocument();
  expect(screen.queryByText("airport 2")).toBeInTheDocument();
  expect(screen.queryByText("airport 3")).not.toBeInTheDocument();
});

test("Test data search for 'airport'", async () => {
  window.scrollTo = jest.fn(); // prevents jest error for unsupported function

  render(<AirportSelect airports={test_airports} setSelection={doNothing} />);

  // Test that the provided elements are available for selection
  fireEvent.click(screen.getByPlaceholderText("Select"));
  expect(screen.queryByText("airport 1")).toBeInTheDocument();
  expect(screen.queryByText("airport 2")).toBeInTheDocument();
  expect(screen.queryByText("airport 3")).toBeInTheDocument();

  // Test search for "airport"
  fireEvent.change(screen.getByPlaceholderText("Search..."), {
    target: { value: "airport" },
  });
  expect(screen.queryByText("airport 1")).toBeInTheDocument();
  expect(screen.queryByText("airport 2")).toBeInTheDocument();
  expect(screen.queryByText("airport 3")).toBeInTheDocument();
});

test("Test data search for 'asdf'", async () => {
  window.scrollTo = jest.fn(); // prevents jest error for unsupported function

  render(<AirportSelect airports={test_airports} setSelection={doNothing} />);

  // Test that the provided elements are available for selection
  fireEvent.click(screen.getByPlaceholderText("Select"));
  expect(screen.queryByText("airport 1")).toBeInTheDocument();
  expect(screen.queryByText("airport 2")).toBeInTheDocument();
  expect(screen.queryByText("airport 3")).toBeInTheDocument();

  // Test search for "asdf"
  fireEvent.change(screen.getByPlaceholderText("Search..."), {
    target: { value: "asdf" },
  });
  expect(screen.queryByText("airport 1")).not.toBeInTheDocument();
  expect(screen.queryByText("airport 2")).not.toBeInTheDocument();
  expect(screen.queryByText("airport 3")).not.toBeInTheDocument();
});
