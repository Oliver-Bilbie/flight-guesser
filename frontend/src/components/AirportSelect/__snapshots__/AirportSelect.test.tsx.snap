// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
<div
  className="StyledBox-sc-13pk1d4-0 iNDmPr"
>
  <span
    className="StyledText-sc-1sadyjn-0 hpCRgD"
  >
    Destination:
  </span>
  <div
    className="StyledBox__StyledBoxGap-sc-13pk1d4-1 qdAYv"
  />
  <div
    className="StyledBox-sc-13pk1d4-0 bsRcKz"
  >
    <div
      aria-label="loading"
      className="StyledBox-sc-13pk1d4-0 dSgfBA"
    />
  </div>
</div>
`;
